ospf6d/ospf6_asbr.c

```diff
@@ -2153,25 +2156,65 @@ static const struct route_map_rule_cmd
 	ospf6_routemap_rule_set_metric_type_free,
 };
 
+struct ospf6_metric {
+	enum { metric_increment, metric_decrement, metric_absolute } type;
+	bool used;
+	uint32_t metric;
+};
+
 static enum route_map_cmd_result_t
 ospf6_routemap_rule_set_metric(void *rule, const struct prefix *prefix,
 			       void *object)
 {
-	char *metric = rule;
-	struct ospf6_route *route = object;
+	struct ospf6_metric *metric;
+	struct ospf6_route *route;
+
+	/* Fetch routemap's rule information. */
+	metric = rule;
+	route = object;
+
+	/* Set metric out value. */
+	if (!metric->used)
+		return RMAP_OKAY;
+
+	route->path.cost = route->path.redistribute_cost;
+
+	if (metric->type == metric_increment)
+		route->path.cost += metric->metric;
+	else if (metric->type == metric_decrement)
+		route->path.cost -= metric->metric;
+	else if (metric->type == metric_absolute)
```

Comment:
Don't you need range checking for the increment and decrement cases?

Reply:
I added range checking in those cases.
As this code is basically copy-paste from ospfd, I think there is also a problem - except if the range checking is done somewhere else.